Source/Utils/Inc/Utils/Handle/Handle.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2024 Steven Frost and Orion Lyau
+
+#pragma once
+
+#include <cstdint>
+#include <limits>
+#include <utility>
+
+// -----------------------------------------------------------------------------
+
+namespace Utils
+{
+
+// -----------------------------------------------------------------------------
+
+template< class TId >
+class Handle
+{
+public:
+
+	static_assert( std::is_integral< TId >(), "TId must be an integral type" );
+
+	static Handle Make()
+	{
+		if ( NextId == std::numeric_limits< TId >::max() )
+		{
+			return Handle();
+		}
+
+		return Handle( NextId++ );
+	}
+
+	Handle()
+		: Id( InvalidId )
+	{}
+
+	explicit Handle( const TId Id )
+		: Id( Id )
+	{}
+
+	Handle( const Handle& Other ) = default;
+	Handle( Handle&& Other ) = default;
+
+	Handle& operator=( const Handle& Other ) = default;
+	Handle& operator=( Handle&& Other ) = default;
+
+	bool operator==( const Handle& Other ) const
+	{
+		return Id == Other.Id;
+	}
+
+	bool operator!=( const Handle& Other ) const
+	{
+		return !( *this == Other );
+	}
+
+	TId GetId() const
+	{
+		return Id;
+	}
+
+	bool IsValid() const
+	{
+		return Id != InvalidId;
+	}
+
+	void Reset()
+	{
+		Id = InvalidId;
+	}
+
+private:
+
+	static TId	InvalidId;
+	static TId	NextId;
+
+	TId			Id;
+};
+
+// -----------------------------------------------------------------------------
+
+template< class TId > TId Handle< TId >::InvalidId = 0;
+template< class TId > TId Handle< TId >::NextId = 1;
```

Comment:
If `NextId` starts at 1, and `TId` is signed (e.g. in the above `typedef Handle< int32_t > EventHandle;`), do we effectively lose half the possible values?

Reply:
Yeah, going to rework this a little so we use the full range of the type & allow customisation if needed by specific types.